packages/graph/hash_graph/lib/graph/src/store/postgres/query/compile.rs

```diff
@@ -322,42 +322,70 @@ impl<'f: 'q, 'q, T: PostgresQueryRecord<'q>> SelectCompiler<'f, 'q, T> {
         }
     }
 
-    /// Joins the list of [`Table`]s as [`JoinExpression`]s.
+    /// Joins a chain of [`Relation`]s and returns the table name of the last joined table.
     ///
     /// Joining the tables attempts to deduplicate [`JoinExpression`]s. As soon as a new filter was
     /// compiled, each subsequent call will result in a new join-chain.
-    fn add_join_statements(
-        &mut self,
-        tables: impl IntoIterator<Item = (TableName, EdgeJoinDirection)>,
-    ) -> Table {
+    fn add_join_statements(&mut self, tables: impl IntoIterator<Item = Relation<'q>>) -> Table {
         let mut current_table = T::base_table();
-        let mut current_edge_direction = EdgeJoinDirection::SourceOnTarget;
-        for (table_name, edge_direction) in tables {
-            if table_name == T::base_table().name && self.artifacts.current_alias.chain_depth == 0 {
-                // Avoid joining the same initial table
-                current_edge_direction = edge_direction;
-                continue;
-            }
-
-            let table = Table {
-                name: table_name,
+        for Relation {
+            current_column_access,
+            join_table_name,
+            join_column_access,
+        } in tables
+        {
+            let current_column = Column {
+                table: current_table,
+                access: current_column_access,
+            };
+            let join_table = Table {
+                name: join_table_name,
                 alias: Some(self.artifacts.current_alias),
             };
+            let join_column = Column {
+                table: join_table,
+                access: join_column_access,
+            };
+
+            // If we join on the same column as the previous join, we can reuse the that join. For
+            // example, if we join on `entities.entity_type_version_id = entity_type.version_id` and
+            // then on `entity_type.version_id = type_ids.version_id`, we can merge the two joins
+            // into `entities.entity_type_version_id = type_ids.version_id`. We, however, need to
+            // make to make sure, that we only alter a join we added in this iteration.
+            if self.artifacts.current_alias.chain_depth != 0 {
+                // Check if we are joining on the same column as the previous join
+                if let Some(last_join) = self.statement.joins.last_mut() {
+                    if last_join.join == current_column {
+                        last_join.join.table.name = join_column.table.name;
+                        last_join.join.access = join_column.access;
+                        current_table = last_join.join.table;
 
-            let join = JoinExpression::from_tables(table, current_table, current_edge_direction);
+                        if let [.., previous_join, this_join] = self.statement.joins.as_slice() {
+                            // It's possible that we just duplicated the last two join statements,
```

Comment:
So here `this_join` would be the `last_join` from above and the `previous_join` is the one before that, which could have been modified to be identical, is that correctly understood from my end? :)